Review and update Interface doc

lread opened the pull request:
Add section headers.

Edits for clarity.

Some of the tips at the end of the doc were very confusing to me. After asking for clarification on Slack, rewrote some of them.

lread committed:
Review and update Interface doc

Add section headers.

Edits for clarity.

Some of the tips at the end of the doc were very confusing to me.
After asking for clarification on Slack, rewrote some of them.

tengstrand commented on doc/interface.adoc:
```diff
@@ -45,16 +49,19 @@ To give an example, let's pretend we have the interface namespace `user.interfac
-Now we are free to edit the `interface.clj` file for both `user` and `admin`, which means they can get out of sync if we are not careful enough.
-Luckily, the Polylith tool will help us keep them consistent, and complain if they differ when we run the xref:commands.adoc#check[check], xref:commands.adoc#info[info] or xref:commands.adoc#test[test] commands.
+Because we are free to edit the `interface.clj` file for both user and admin, they can get out of sync if we are not careful. 
+Luckily, the Polylith tool will help us keep them consistent and complain if they differ when we run the xref:commands.adoc#check[check], xref:commands.adoc#info[info] or xref:commands.adoc#test[test] commands.
```
Her should use "the poly `tool`" (where poly is marked as code, which is also a convention through out the documentation). In the high-level doc we sometimes use "the Polylith tool" because that documentation is generic, but here we know that we talk about the `poly` tool. I missed to change this, so this was not introduced by you.

lread replied:
Yeah, that's a good convention. I'll update.

tengstrand commented on doc/interface.adoc:
```diff
@@ -126,46 +134,61 @@ There is no magic here, just include the definitions you want, like this:
+The implementing function can expose specific usage via destructuring, e.g., `sell [{:keys [model type color]}]`.
 
-* If using http://clojure-doc.org/articles/language/functions.html#variadic-functions[variadic functions] in the interface, a simplification is to pass in what comes after `&` as a vector to the implementing function.
+* It sometimes makes sense for a http://clojure-doc.org/articles/language/functions.html#multi-arity-functions[multi-arity function] in an interface to delegate to a single arity function in the implementing namespace. 
```
End the sentence with a colon (the same applies to the next point).

lread replied:
Can do

tengstrand commented on doc/interface.adoc:
```diff
@@ -126,46 +134,61 @@ There is no magic here, just include the definitions you want, like this:
-* We want to share code between Clojure and ClojureScript via `cljc` files.
-Since `interface` is a reserved word in ClojureScript, it could otherwise cause problems.
+* Polylith will always recognize `interface` and `ifc` as interface namespace names.
+By default, it will generate code using `interface` as the interface namespace name.
```
Add: "when you create a component." in the end

lread replied:
Can do

tengstrand commented on doc/interface.adoc:
```diff
@@ -126,46 +134,61 @@ There is no magic here, just include the definitions you want, like this:
+Since `interface` is a reserved word in ClojureScript, it will cause problems.
+In this case, you can either:
+*** set `:interface-ns` to `ifc`, `poly` will use `ifc` as the interface namespace name for all generated code
+*** or leave `interface` as the default and manually rename any generated `interface` to `ifc` for the components that will also run from ClojureScript.
```
We don't have to manually change the namespace here. Instead, we can give `interface:ifc` to the `create component` command, and it we use `ifc` instead as the interface name.

lread replied:
Good to know, will add.

tengstrand reviewed (changes requested):
Great improvements! They will definitely make the text easier to understand for users new to Polylith.

tengstrand commented on doc/interface.adoc:
```diff
@@ -126,46 +134,61 @@ There is no magic here, just include the definitions you want, like this:
+* If the interface `foo` occurs in xref:#one-interface-in-multiple-components[multiple components], you cannot define `some-call` as a function in one `foo` interface and a macro in other `foo` interfaces. 
+If `some-call` is a function in one `foo` interface, it must be a function in all `foo` interfaces. If `some-call` is a macro in one `foo` interface, it must be a macro in all `foo` interfaces.
+We have this restriction because functions are composable, but macros are not.
 
```
Here somewhere, we could mention that the `poly` tool also checks that the signature must match exactly, for functions and macros. So if the parameter list is `[a b c]` for function `foo` it has to be defined as `[a b c]` in `foo` for all the implementing components. It means that `[b a c]` is not accepted. Type hints are also checked, e.g. if a function is defined as `[^String x y]`, we have define it as `[^String x y]` everywhere for that function. This reduces the risk that we enter the parameters in the wrong order.

lread replied:
These matching requirements are described above under the "Declaring the Same Interface in Multiple Components" section. I think it would be clearer if this whole tip was moved up to that section where it would be in context. 

I also like your explicit function signature examples, I'll work them in.
I expect I could make `defn` vs `defmacro` less wordy and clearer with examples.

lread committed:
review feedback: Polylith tool -> `poly` tool

lread committed:
review feedback: use : instead of .

lread committed:
review feedback: tip on interface-ns

Clarify that generated code is from creating a component.

Mention `create component` `interface:` option.

lread committed:
Rework "same interface" mismatch check docs

Move tip about functions and macros into appropriate section.
Rework text into table.
Content is now succinct and easy to follow.

lread committed:
styling tweak: avoid wrapping examples in table

tengstrand commented on doc/interface.adoc:
```diff
@@ -45,16 +49,42 @@ To give an example, let's pretend we have the interface namespace `user.interfac
+Because we are free to edit the `interface.clj` file for both user and admin, they can get out of sync if we are not careful. 
+Luckily, the `poly` tool will help us keep them consistent and complain if they differ when we run the xref:commands.adoc#check[check], xref:commands.adoc#info[info] or xref:commands.adoc#test[test] commands:
+
+[cols="25,25,50"]
```
I really like this change of introducing a table here!

lread replied:
Glad you like it too!

tengstrand commented on doc/interface.adoc:
```diff
@@ -45,16 +49,42 @@ To give an example, let's pretend we have the interface namespace `user.interfac
+To do so, create an `interface` package (directory) with the name `interface` at the component root and then add the sub-namespaces.
+
+You can find an example in Polylith itself.
+The `util` component divides its https://github.com/polyfy/polylith/tree/master/components/util/src/polylith/clj/core/util/interface[interface] into several sub-namespaces:
```
Here I prefer one sentence, because they are so coupled to each other:
You can find an example in Polylith itself, where the `util` component divides its...

lread replied:
Yeah, I think we'll continue to have a healthy back-and-forth on sentence length.
I see no problem with joining here.

tengstrand commented on doc/interface.adoc:
```diff
@@ -67,10 +97,11 @@ util
+You can see an example of this in the https://github.com/furkan3ayraktar/clojure-polylith-realworld-example-app[RealWorld example app], where the `article` component has both an `article.interface` and an https://github.com/furkan3ayraktar/clojure-polylith-realworld-example-app/blob/master/components/article/src/clojure/realworld/article/interface/spec.clj[`article.interface.spec`] sub-interface.
 
-It can then be used from e.g. the https://github.com/furkan3ayraktar/clojure-polylith-realworld-example-app/blob/master/bases/rest-api/src/clojure/realworld/rest_api/handler.clj[handler] namespace in `rest-api`:
+Here's an example of the sub-interface usage from the https://github.com/furkan3ayraktar/clojure-polylith-realworld-example-app/blob/master/bases/rest-api/src/clojure/realworld/rest_api/handler.clj[RealWorld example app handler] namespace in its `rest-api`:
```
End the sentence with "base" (that links to the `base` page):
...in it's `rest-api` base:

lread replied:
Can do

tengstrand commented on doc/interface.adoc:
```diff
@@ -126,46 +157,57 @@ There is no magic here, just include the definitions you want, like this:
 
-* If using http://clojure-doc.org/articles/language/functions.html#variadic-functions[variadic functions] in the interface, a simplification is to pass in what comes after `&` as a vector to the implementing function.
+* The interface can expose the entity's name, e.g., `sell [car]`. 
+The implementing function can expose specific usage via destructuring, e.g., `sell [{:keys [model type color]}]`.
```
Here the comparison is lost when we split the sentence in two. I recommend to put back "while" again, to make it clearer that they have two different roles.

lread replied:
I find it clearer with 2 sentences.
They are in the same paragraph, after all.
But I think we are on subjective grounds here maybe? So also fine with joining them.
If only we had an English teacher to chime in! :slightly_smiling_face:

lread replied:
Hmm... maybe you are right. That does seem to read better!

lread committed:
review feedback: merge 2 short sentences

lread committed:
review feedback: mention/link to base docs

lread committed:
review feedback: join 2 sentences
